cmd/ddev/cmd/a.go

```diff
@@ -2,19 +2,82 @@ package cmd
 
 import (
 	"os"
+	"strings"
 
+	"github.com/ddev/ddev/pkg/ddevapp"
 	"github.com/ddev/ddev/pkg/dockerutil"
 	"github.com/ddev/ddev/pkg/globalconfig"
+	"github.com/ddev/ddev/pkg/output"
+	"github.com/spf13/pflag"
 )
 
-// This file is a.go because global config must be loaded before anybody else
-// runs their init(), as they might overwrite global_config.yaml with
-// uninitialized data
+// This file is a.go because some code needs to be processed before any other
+// files run their init() to avoid conflicts and ensure certain functionality
+// is available.
 
 func init() {
+	// Set up logger
+	output.LogSetUp()
+
+	// Set up global config
 	globalconfig.EnsureGlobalConfig()
 	_ = os.Setenv("DOCKER_CLI_HINTS", "false")
 	_ = os.Setenv("MUTAGEN_DATA_DIRECTORY", globalconfig.GetMutagenDataDirectory())
+
 	// GetDockerClient should be called early to get DOCKER_HOST set
 	_, _ = dockerutil.GetDockerClient()
+
+	// Parse, and use the "--project" flag
+	handleProjectFlag()
+}
+
+// handleProjectFlag parses the "--project" flag, and sets the current working
+// directory to the named project, if any.
+func handleProjectFlag() {
+	// We need a new flag set so we can parse this early without affecting flags that belong to other commands.
+	tempFlags := pflag.NewFlagSet("", pflag.ContinueOnError)
+	// Don't throw errors for unknown flags - we don't know ANY of the flags at this stage except the ones
+	// explicitly added below.
+	tempFlags.ParseErrorsWhitelist = pflag.ParseErrorsWhitelist{UnknownFlags: true}
+	// We need a dummy help flag to avoid pflag.ErrHelp
+	tempFlags.BoolP("help", "h", false, "")
+	tempFlags.String("project", "", "")
+
+	// Parse the flag
+	args := os.Args[1:]
+	err := tempFlags.Parse(args)
+	if err != nil {
+		output.UserErr.Fatal("Couldn't parse --project flag: ", err)
+	}
+	project, err := tempFlags.GetString("project")
+	if err != nil {
+		// This shouldn't ever happen - but go wants us to handle the error.
+		output.UserErr.Fatal("Unexpected error getting value from --project flag: ", err)
+	}
+
+	if project != "" {
+		// Find and change to the root dir for the project
+		approot, err := ddevapp.GetActiveAppRoot(project)
+		if err != nil {
+			output.UserErr.Fatal("Couldn't find project: ", err)
+		}
```

Comment:
So... there's a trade-off to this line.

On the one hand, this protects against running commands somewhere you didn't expect to. Clearly if you included the flag then you intended to run the command in a specific project, so if we just continued even though we can't swap to whatever project you wanted that would be bad.

On the other hand, this breaks completion for the project flag itself, e.g. `ddev --project=m` will _not_ give you any valid project names as suggestions, because execution of the completion hits this code and exits as soon as it can't find a project named "m".

I'm gonna play a bit more and see if I can find a clean way to detect (at this very early stage) that it's a completion request, but we might just have to accept that completion for this flag is a little wonky.

Reply:
And of course I immediately found a solution 😂
Feels just about as janky as https://github.com/ddev/ddev/pull/5863#discussion_r1497062919 but at the very least it shows this is possible.

Reply:
This is mostly about seeing what's even possible at this stage. When I have something that works without breaking CI then I can look at making it more robust. I'm mostly exploring the problem space right now.